File: MyFarmProduce/Services/CartFactory.cs
using MyFarmProduce.Application.Interfaces;
using MyFarmProduce.Common;
using MyFarmProduce.Web.Models;

namespace MyFarmProduce.Web.Services;

/// <summary>Builds a display cart by joining session quantities with current product data.</summary>
public class CartFactory
{
    private readonly ICartService _cart;
    private readonly ICatalogService _catalog;

    public CartFactory(ICartService cart, ICatalogService catalog)
    {
        _cart = cart;
        _catalog = catalog;
    }

    public async Task<CartViewModel> BuildAsync()
    {
        var items = _cart.GetItems();
        var vm = new CartViewModel();

        if (items.Count > 0)
        {
            var products = await _catalog.GetProductsByIdsAsync(items.Keys);
            foreach (var product in products)
            {
                vm.Lines.Add(new CartLineViewModel
                {
                    Product = product,
                    Quantity = items[product.Id]
                });
            }
            vm.Lines = vm.Lines.OrderBy(l => l.Product.Name).ToList();
        }

        vm.DeliveryFee = vm.IsEmpty ? 0m : AppConstants.FlatDeliveryFee;
        return vm;
    }
}

File: web/src/main/java/com/myfarmproduce/web/service/CartFactory.java
package com.myfarmproduce.web.service;

import com.myfarmproduce.application.service.CatalogService;
import com.myfarmproduce.common.AppConstants;
import com.myfarmproduce.domain.entity.Product;
import com.myfarmproduce.web.dto.CartLineViewModel;
import com.myfarmproduce.web.dto.CartViewModel;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.Map;

/** Builds a display cart by joining session quantities with current product data. */
@Component
public class CartFactory {

    private final CartService cart;
    private final CatalogService catalog;

    public CartFactory(CartService cart, CatalogService catalog) {
        this.cart = cart;
        this.catalog = catalog;
    }

    public CartViewModel build() {
        Map<Integer, Integer> items = cart.getItems();
        CartViewModel vm = new CartViewModel();

        if (!items.isEmpty()) {
            for (Product product : catalog.getProductsByIds(items.keySet())) {
                CartLineViewModel line = new CartLineViewModel();
                line.setProduct(product);
                line.setQuantity(items.get(product.getId()));
                vm.getLines().add(line);
            }
            vm.setLines(vm.getLines().stream()
                    .sorted(Comparator.comparing(l -> l.getProduct().getName()))
                    .toList());
        }

        vm.setDeliveryFee(vm.isEmpty() ? BigDecimal.ZERO : AppConstants.FLAT_DELIVERY_FEE);
        return vm;
    }
}
